### DISCUSSION TITLE
SharpPcap tunnel device linux

### DISCUSSION THREAD

**Discussion by trudyhood** [chmorgan/sharppcap#476] [Q&A] [answered]
I'm currently involved in a VPN project utilizing SharpPCAP on the client side, which can be found here: https://github.com/vpnhood/VpnHood.

I'm keen to explore the possibility of using SharpPCAP for the server side as well. I've observed that the TunnelDevice.GetTunnelInterfaces method returns an empty array when running it on a Linux virtual machine (VM). I'm wondering if SharpPCAP TunnelDevice is compatible with Linux, as I couldn't locate any examples or documentation showcasing the use of SharpPCAP with TunnelDevice on Linux. Any guidance or insights on this matter would be greatly appreciated.

**Comment by kayoub5** [comment, marked as answer]
Hi,

It is tested to be working on ubuntu, you will need to create the tuntap device using `ip tuntap add mode tap`. 

Please take into account that unlike Windows, tuntap requires sudo access by default on Linux.

**Reply by trudyhood** [reply]
Thank you so much. This is what I was looking for.

```
ip tuntap add mode tap
```

This is unbelievably simple with SharpPCAP to start sending and receiving packets as this:

```c#
            var nic = TunnelDevice.GetTunnelInterfaces().Skip(1).First();
            TunnelDevice tunnelDevice = new TunnelDevice(nic);
            tunnelDevice.Open();
            tunnelDevice.OnPacketArrival+= (sender, e) =>
            {
                Console.WriteLine("Packet arrived:" + e.GetPacket());
            };
            tunnelDevice.StartCapture();
            var packet = CreateUdpPacket(IPEndPoint.Parse("8.8.8.8:51"), IPEndPoint.Parse("127.0.0.0:920"), Array.Empty<byte>());
            tunnelDevice.SendPacket(packet);
```

**Reply by trudyhood** [reply]
> Hi,
> 
> It is tested to be working on ubuntu, you will need to create the tuntap device using `ip tuntap add mode tap`.
> 
> Please take into account that unlike Windows, tuntap requires sudo access by default on Linux.

Does this work on TUN device too or just TAP?

**Reply by kayoub5** [reply]
TAP only, adding TUN support is possible, but I don't have time for it, PR is welcome
